llvm/lib/Target/BPF/BPFInstrInfo.td

```diff
@@ -806,6 +806,7 @@ class XADD<BPFWidthModifer SizeOp, string OpcodeStr, PatFrag OpNode>
 let Constraints = "$dst = $val" in {
   let Predicates = [BPFNoALU32] in {
     def XADDW : XADD<BPF_W, "u32", atomic_load_add_i32>;
+    def XADDD : XADD<BPF_DW, "u64", atomic_load_add_i64>;
```

Comment:
Would it be possible to get rid of `XADD` in favour of `ATOMIC[32]_NOFETCH` and a few `def : Pat`? I tried to do so [remove-xadd.txt](https://github.com/user-attachments/files/16807144/remove-xadd.txt), but that is not enough.

Reply:
Not sure about this. Let me give a try.

Reply:
@eddyz87 I checked your code. It won't work because with BPFNoALU32 does not allow sub-registers. In current implementation, 64bit atomic_nofetch insns are available at cpu v1, but 32bit atomic nofetch (except xadd) are not available at cpu v1. so 32bit xadd needs specialization anyway. so your approach will work for 64bit xaddd, but not for xadd. So I suggest to use the current implementation. Let me know what you think.

Reply:
Right, that thing was not complete and I couldn't figure why, thank you for explaining.
Here is a fixed version:  [remove-xadd.txt](https://github.com/user-attachments/files/16816421/remove-xadd.txt)
What I like here is that all xadd variants are grouped together, but up to you.

Reply:
@eddyz87 What you suggested indeed simpler and I just incorporated it into the latest revision. I made two changes:
  - Remove 'let mayStore = 1' in ATOMIC_NOFETCH. I am not whether this is needed or not. Previous class does not have this.
  - Remove 'MI.getOpcode() != BPF::XADDW32' in BPFMIChecking.cpp. XADDW32 is only available with alu32, in which case return value is okay.

Please help take a look again in case that I missed anything. Thanks!

Reply:
> Remove 'let mayStore = 1' in ATOMIC_NOFETCH. I am not whether this is needed or not. Previous class does not have this.

I think tablegen wanted me to add this flag when I added patterns. If it does not report warnings on you side, then ok.

> Remove 'MI.getOpcode() != BPF::XADDW32' in BPFMIChecking.cpp. XADDW32 is only available with alu32, in which case return value is okay.

`XADDW32` is not a member of any pattern, so I agree, this should be fine.

Reply:
For
>  I think tablegen wanted me to add this flag when I added patterns. If it does not report warnings on your side, then ok.

No, I didn't see any warnings.

Reply:
Ok, all good then